kafka-rest/src/main/java/io/confluent/kafkarest/resources/v3/ProduceRateLimiters.java

```diff
@@ -20,66 +20,79 @@
+import io.confluent.kafkarest.ratelimit.RequestRateLimiter;
+import io.confluent.kafkarest.ratelimit.RequestRateLimiterProduceBytes;
+import io.confluent.kafkarest.ratelimit.RequestRateLimiterProduceCount;
 import java.time.Duration;
 import java.util.Optional;
 import java.util.concurrent.TimeUnit;
 import javax.inject.Inject;
+import javax.inject.Provider;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 public class ProduceRateLimiters {
 
-  private final int maxRequestsPerSecond;
-  private final int maxBytesPerSecond;
-  private final Duration gracePeriod;
   private final boolean rateLimitingEnabled;
-  private final LoadingCache<String, ProduceRateLimiter> cache;
+  private final LoadingCache<String, RequestRateLimiter> countCache;
+  private final LoadingCache<String, RequestRateLimiter> bytesCache;
+  Provider<RequestRateLimiterProduceCount> countLimiterProvider;
+  Provider<RequestRateLimiterProduceBytes> bytesLimiterProvider;
+
+  private static final Logger log = LoggerFactory.getLogger(ProduceRateLimiters.class);
 
   @Inject
   public ProduceRateLimiters(
-      @ProduceGracePeriodConfig Duration produceGracePeriodConfig,
-      @ProduceRateLimitCountConfig Integer produceRateLimitCountConfig,
-      @ProduceRateLimitBytesConfig Integer produceRateLimitBytesConfig,
+      Provider<RequestRateLimiterProduceCount> countLimiterProvider,
+      Provider<RequestRateLimiterProduceBytes> bytesLimiterProvider,
       @ProduceRateLimitEnabledConfig Boolean produceRateLimitEnabledConfig,
-      @ProduceRateLimitCacheExpiryConfig Duration produceRateLimitCacheExpiryConfig,
-      Clock time) {
-    this.maxRequestsPerSecond = requireNonNull(produceRateLimitCountConfig);
-    this.gracePeriod = requireNonNull(produceGracePeriodConfig);
+      @ProduceRateLimitCacheExpiryConfig Duration produceRateLimitCacheExpiryConfig) {
     this.rateLimitingEnabled = requireNonNull(produceRateLimitEnabledConfig);
-    this.maxBytesPerSecond = requireNonNull(produceRateLimitBytesConfig);
-    requireNonNull(time);
+    this.countLimiterProvider = countLimiterProvider;
+    this.bytesLimiterProvider = bytesLimiterProvider;
 
-    cache =
+    countCache =
         CacheBuilder.newBuilder()
             .expireAfterAccess(produceRateLimitCacheExpiryConfig.toMillis(), TimeUnit.MILLISECONDS)
             .build(
-                new CacheLoader<String, ProduceRateLimiter>() {
+                new CacheLoader<String, RequestRateLimiter>() {
                   @Override
-                  public ProduceRateLimiter load(String key) {
-                    return new ProduceRateLimiter(
-                        gracePeriod,
-                        maxRequestsPerSecond,
-                        maxBytesPerSecond,
-                        produceRateLimitEnabledConfig,
-                        time);
+                  public RequestRateLimiter load(String key) {
+                    return countLimiterProvider.get().getLimiter();
```

Comment:
You don't need `getLimiter`, and this class does not need to be `public`. Use binding annotations instead to differentiate between all the `RequestRateLimiters`.